File: exams/collahuazo/u3/Strategy/sorting_strategy_app/controller/sort_controller.py
from controller.sorting_context import SortingContext
from controller.bubble_sort import BubbleSort
from controller.insertion_sort import InsertionSort
from controller.quick_sort import QuickSort
from model.sort_record import SortRecord
from utils.db_connection import DBConnection


class SortController:
    def __init__(self):
        self.context = SortingContext()
        self.db = DBConnection()

    def parse_input(self, text: str) -> list:
        parts = text.split(",")
        numbers = []

        for part in parts:
            part = part.strip()

            if part == "":
                continue

            try:
                number = int(part)
            except ValueError:
                raise ValueError("All elements must be separated by comas")

            numbers.append(number)

        if len(numbers) == 0:
            raise ValueError("Enter one number.")

        return numbers

    def select_algorithm(self, size: int) -> str:
        if size < 2:
            raise ValueError("The array must have more than one element")
        elif size >= 2 and size <= 6:
            self.context.setSortStrategy(BubbleSort())
            return "BubbleSort"
        elif size >= 7 and size <= 10:
            self.context.setSortStrategy(InsertionSort())
            return "InsertionSort"
        else:
            self.context.setSortStrategy(QuickSort())
            return "QuickSort"

    def numbers_to_text(self, numbers: list) -> str:
        text = ""
        for i in range(len(numbers)):
            text = text + str(numbers[i])
            if i < len(numbers) - 1:
                text = text + ", "
        return text

    def process(self, text: str) -> SortRecord:
        numbers = self.parse_input(text)
        size = len(numbers)

        algorithm_name = self.select_algorithm(size)
        sorted_numbers = self.context.sort(numbers)

        unsorted_str = self.numbers_to_text(numbers)
        sorted_str = self.numbers_to_text(sorted_numbers)

        record = SortRecord(unsorted_str, size, algorithm_name, sorted_str)

        inserted_id = self.db.insert_record(record.to_dict())
        record.saved = inserted_id is not None

        return record

File: exams/palomo/u3/ExamP3/Strategy/src/ec/edu/espe/Strategy/controller/SortController.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */

package ec.edu.espe.Strategy.controller;

import ec.edu.espe.Strategy.utils.MongoDBConnection;
import ec.edu.espe.Strategy.model.SortingContext;
import ec.edu.espe.Strategy.model.SortingRecord;
import ec.edu.espe.Strategy.model.SortingStrategy;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 *
 * @author Cristian Palomo, Error 404, @ESPE
 */
public class SortController {
    
    private final SortingContext sortingContext;

    public SortController() {
        this.sortingContext = new SortingContext();
    }

    public SortingRecord processAndSort(String inputString) throws IllegalArgumentException {
        int[] unsortedArray = parseAndValidateInput(inputString);
        int size = unsortedArray.length;

        if (size <= 1) {
            throw new IllegalArgumentException("The number of elements must be greater than 1 (Entered size: " + size + ").");
        }

        // Formateamos el texto del arreglo desordenado
        String unsortedStr = formatArrayToString(unsortedArray);

        // Seleccionamos la estrategia y ejecutamos el ordenamiento
        SortingStrategy selectedStrategy = sortingContext.setSortStrategyBasedOnSize(size);
        String algorithmName = selectedStrategy.getClass().getSimpleName();
        
        // Enviamos una copia para evitar mutar el arreglo original si se ordena in-place
        int[] sortedArray = sortingContext.sort(unsortedArray.clone());
        String sortedStr = formatArrayToString(sortedArray);

        SortingRecord record = new SortingRecord(unsortedStr, size, algorithmName, sortedStr);

        // Guardado en la base de datos
        if (!MongoDBConnection.saveSortingRecord(record)) {
            System.err.println("Warning: Result was generated locally but could not be saved to MongoDB.");
        }

        return record;
    }

    public List<SortingRecord> fetchHistory() {
        return MongoDBConnection.getAllRecords();
    }

    // --- Métodos de apoyo (Helper Methods) ---

    private int[] parseAndValidateInput(String inputString) {
        if (inputString == null || inputString.trim().isEmpty()) {
            throw new IllegalArgumentException("Please enter numeric elements separated by commas.");
        }

        List<Integer> list = new ArrayList<>();
        for (String token : inputString.split(",")) {
            String trimmed = token.trim();
            if (!trimmed.isEmpty()) {
                try {
                    list.add(Integer.parseInt(trimmed));
                } catch (NumberFormatException e) {
                    throw new IllegalArgumentException("Invalid element: '" + trimmed + "'. Please enter integers only.");
                }
            }
        }

        return list.stream().mapToInt(Integer::intValue).toArray();
    }

    private String formatArrayToString(int[] array) {
        return Arrays.toString(array).replaceAll("[\\[\\]]", "");
    }
}

